Display formatting support (#584)

dalance opened the pull request:
This PR adds display formatting support.

Closes #584.

ordian commented:
Thank you for the PR!
We're using `cargo +nightly fmt` with the latest nightly: https://github.com/paritytech/parity-common/blob/5fe7924049f39fc6cf772605efc5c28b3ecbae78/.github/workflows/ci.yml#L157

ordian commented on uint/tests/uint_tests.rs:
```diff
@@ -540,6 +540,19 @@ fn display_uint_zero() {
 
+#[test]
+fn display_uint_format() {
+	let s = "12345678987654321023456789";
```
could you also test formatting 0 since we're special casing it?
```suggestion
	let s = U256::from_dec_str("12345678987654321023456789").unwrap();
```

ordian commented on uint/src/uint.rs:
```diff
@@ -1673,22 +1667,46 @@ macro_rules! construct_uint {
 				}
+				f.pad_integral(true, "0x", &buf)
+			}
+		}
```
can we reuse the code in LowerHex and UpperHex?

ordian commented:
Some code duplication nits, otherwise, looks good.

dalance commented:
I fixed code duplication and added test for 0.
I tried to `cargo +nightly fmt`, but there are some format changings yet.
Now I use `rustfmt 1.4.38-nightly (db9d361 2021-11-28)`.

dalance commented:
Sorry. I refactored 0 display test.

ordian commented on uint/src/uint.rs:
```diff
@@ -1307,6 +1301,34 @@ macro_rules! construct_uint {
+				}
+
+				let mut latch = false;
+				let mut buf = String::new();
```
hmm, this doesn't compile for no_std
why do we use a string and not directly write to f as before?

dalance replied:
This is because `pad_integral` for padding requires constructed `str`.
If I implement `pad_integral` manually, `String` can be removed probably.
I'll try it.

dalance commented:
I removed `String` by using fixed size array as the same as `Display` implementation.

ordian commented on uint/src/uint.rs:
```diff
@@ -1307,6 +1301,47 @@ macro_rules! construct_uint {
+						if latch {
+							let nibble = match nibble {
+								0..=9 => nibble as u8 + b'0',
+								10 => if is_lower { b'a' } else { b'A'},
+								11 => if is_lower { b'b' } else { b'B'},
+								12 => if is_lower { b'c' } else { b'C'},
+								13 => if is_lower { b'd' } else { b'D'},
+								14 => if is_lower { b'e' } else { b'E'},
+								15 => if is_lower { b'f' } else { b'F'},
+								_ => unreachable!(),
```
```suggestion
								_ if is_lower => nibble as u8 + b'a',
								_ => nibble as u8 + b'A',
```

dalance commented:
I applied the suggested code.

dalance commented:
I fixed one more thing.
If there is not nibble range check by `unreachable`, `from_utf8_unchecked` may be unsound.

ordian commented:
> If there is not nibble range check by unreachable, from_utf8_unchecked may be unsound.

nibble is guaranteed to be in the 0..=15 range by construction, I don't think we need the unreachable check, which is why I suggested to remove it in the code suggestion.

dalance commented:
> nibble is guaranteed to be in the 0..=15 range by construction, I don't think we need the unreachable check, which is why I suggested to remove it in the code suggestion.

I understand nibble < 16, but it is not obvious from code.
How about the following code?

```rust
let nibble = (ch & (15u64 << ((15 - x) * 4) as u64)) >> (((15 - x) * 4) as u64);
assert!(nibble < 16);
```

ordian commented:
`debug_assert` might be more appropriate, I'm worried about premature pessimization of code

dalance commented:
OK. I added more comments instead of `assert`.

ordian commented:
Thank you.

niklasad1 reviewed (approved):
LGTM

niklasad1 commented on uint/src/uint.rs:
```diff
@@ -1307,6 +1301,44 @@ macro_rules! construct_uint {
+				}
+
+				let mut latch = false;
+				let mut buf = [0_u8; $n_words*16];
```
```suggestion
				let mut buf = [0_u8; $n_words * 16];
```

dalance commented:
Thank you.
Could you bump up version and release to crates.io?
